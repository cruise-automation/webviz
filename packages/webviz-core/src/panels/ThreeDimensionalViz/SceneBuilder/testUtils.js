// @flow
//
//  Copyright (c) 2021-present, Cruise LLC
//
//  This source code is licensed under the Apache License, Version 2.0,
//  found in the LICENSE file in the root directory of this source tree.
//  You may not use this file except in compliance with the License.
import type { Interactive } from "webviz-core/src/panels/ThreeDimensionalViz/Interactions/types";
import type { Marker } from "webviz-core/src/types/Messages";
import type { MarkerCollector } from "webviz-core/src/types/Scene";

export class OpenSourceMarkerCollector implements MarkerCollector {
  markers: Interactive<Marker>[] = [];
  grids: any[] = [];

  arrow = (o: any) => this.markers.push(o);
  cube = (o: any) => this.markers.push(o);
  cubeList = (o: any) => this.markers.push(o);
  sphere = (o: any) => this.markers.push(o);
  cylinder = (o: any) => this.markers.push(o);
  lineStrip = (o: any) => this.markers.push(o);
  lineList = (o: any) => this.markers.push(o);
  sphereList = (o: any) => this.markers.push(o);
  points = (o: any) => this.markers.push(o);
  text = (o: any) => this.markers.push(o);
  triangleList = (o: any) => this.markers.push(o);
  poseMarker = (o: any) => this.markers.push(o);
  filledPolygon = (o: any) => this.markers.push(o);
  instancedLineList = (o: any) => this.markers.push(o);
  overlayIcon = (o: any) => this.markers.push(o);
  grid(o: any) {
    this.grids.push(o);
  }
  pointcloud(o: any) {
    this.markers.push(o);
  }
  laserScan(_o: any) {
    throw new Error("not used in test");
  }
  radarPointCluster(_o: any) {
    throw new Error("not used in test");
  }
  linedConvexHull(_o: any) {
    throw new Error("not used in test");
  }
}
