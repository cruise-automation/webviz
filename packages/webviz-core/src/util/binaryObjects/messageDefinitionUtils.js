// @flow
//
//  Copyright (c) 2020-present, Cruise LLC
//
//  This source code is licensed under the Apache License, Version 2.0,
//  found in the LICENSE file in the root directory of this source tree.
//  You may not use this file except in compliance with the License.

import { sum } from "lodash";
import memoize from "memoize-weak";
import { type RosMsgField } from "rosbag";

import type { RosDatatypes } from "webviz-core/src/types/RosDatatypes";

const primitiveSizes = {
  json: 8,
  string: 8,
  bool: 1,
  int8: 1,
  uint8: 1,
  int16: 2,
  uint16: 2,
  int32: 4,
  uint32: 4,
  float32: 4,
  float64: 8,
  int64: 8,
  uint64: 8,
  time: 8,
  duration: 8,
};

export const primitiveList: Set<string> = new Set(Object.keys(primitiveSizes));

export class PointerExpression {
  variable: string;
  constant: number;

  constructor(variable: string, _constant: number = 0) {
    this.variable = variable;
    this.constant = _constant;
  }
  toString() {
    if (this.constant === 0) {
      return this.variable;
    }
    return `(${this.variable} + ${this.constant})`;
  }
  add(o: number) {
    return new PointerExpression(this.variable, this.constant + o);
  }
}

export const typeSize = memoize(
  (typesByName: RosDatatypes, typeName: string): number => {
    if (primitiveSizes[typeName] != null) {
      return primitiveSizes[typeName];
    }
    const messageType = typesByName[typeName];
    if (messageType == null) {
      throw new Error(`Unknown type: ${typeName}`);
    }
    // We add field sizes here, not type sizes. If the type has a field that's an array of strings,
    // it only adds eight bytes to the object's inline size.
    return sum(typesByName[typeName].fields.map((field) => fieldSize(typesByName, field)));
  }
);

// No point in memoizing this -- it can only do O(1) work before hitting `typeSize`, which is
// memoized. Memoizing on the string type name will work better than memoizing on field object
// identity, too.
export function fieldSize(typesByName: RosDatatypes, field: RosMsgField): number {
  if (field.isConstant) {
    return 0;
  }
  if (field.isArray) {
    // NOTE: Fixed-size arrays are not inlined.
    return 2 * primitiveSizes.int32;
  }
  return typeSize(typesByName, field.type);
}

// It's useful for field-accessor code for times and durations to share codepaths with field
// accessor code for complex message types. To do that, we add them to the set of messages.
// (Note, this might change depending on the code written next.)
export const addTimeTypes = (typesByName: RosDatatypes): RosDatatypes => ({
  ...typesByName,
  time: { fields: [{ name: "sec", type: "int32" }, { name: "nsec", type: "int32" }] },
  duration: { fields: [{ name: "sec", type: "int32" }, { name: "nsec", type: "int32" }] },
});

// String.prototype.replaceAll is not implemented in Chrome.
const allBadCharacters = new RegExp("[/-]", "g");
export const friendlyTypeName = (name: string): string => name.replace(allBadCharacters, "_");
export const deepParseSymbol = Symbol("deepParse");
type CommonBobjectSourceData = $ReadOnly<{|
  datatypes: RosDatatypes,
  datatype: string,
|}>;
export type BobjectSourceData =
  | CommonBobjectSourceData
  | $ReadOnly<{|
      ...CommonBobjectSourceData,
      buffer: ArrayBuffer, // Present in binary bobjects
      bigString: string,
      isArrayView: boolean,
    |}>;
export const classSourceData = new WeakMap<any, BobjectSourceData>();
export const associateSourceData = (cls: any, sourceData: BobjectSourceData): void => {
  classSourceData.set(cls, sourceData);
};
export const getSourceData = (cls: any): ?BobjectSourceData => classSourceData.get(cls);

// Note: returns false for "time" and "duration". Might be more useful not doing that.
// The RosMsgField flow-type has an `isComplex` field, but it isn't present for the websocket
// player, and has gone missing in test fixture before as well. Best to just assume it doesn't
// exist, because it's an annoying denormalization anyway.
export const isComplex = (typeName: string) => !primitiveList.has(typeName);

const numberList = ["int8", "uint8", "int16", "uint16", "int32", "uint32", "float32", "float64", "int64", "uint64"];
export const isNumberType = (typeName: string) => numberList.includes(typeName);
