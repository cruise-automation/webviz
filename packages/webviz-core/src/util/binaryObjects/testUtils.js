// @flow
//
//  Copyright (c) 2020-present, Cruise LLC
//
//  This source code is licensed under the Apache License, Version 2.0,
//  found in the LICENSE file in the root directory of this source tree.
//  You may not use this file except in compliance with the License.

import type { RosValue } from "webviz-core/src/players/types";
import type { BinaryHeader } from "webviz-core/src/types/BinaryMessages";
import type { RosDatatypes } from "webviz-core/src/types/RosDatatypes";
import type { ArrayView } from "webviz-core/src/util/binaryObjects/ArrayViews";

export const definitions: RosDatatypes = {
  "std_msgs/Header": {
    fields: [{ type: "uint32", name: "seq" }, { type: "time", name: "stamp" }, { type: "string", name: "frame_id" }],
  },
  "fake_msgs/HasComplexAndArray": {
    fields: [{ type: "std_msgs/Header", name: "header" }, { type: "string", isArray: true, name: "stringArray" }],
  },
  "fake_msgs/HasComplexArray": {
    fields: [{ type: "fake_msgs/HasComplexAndArray", name: "complexArray", isArray: true }],
  },
  "fake_msgs/HasConstant": {
    fields: [{ type: "uint8", name: "const", isConstant: true, value: 1 }],
  },
  "fake_msgs/HasByteArray": {
    fields: [{ type: "uint8", name: "byte_array", isArray: true }],
  },
  "fake_msgs/HasJson": {
    fields: [{ type: "json", name: "jsonField" }],
  },
  "fake_msgs/HasInt64s": {
    fields: [{ type: "int64", name: "i64" }, { type: "uint64", name: "u64" }],
  },
  "fake_msgs/HasArrayOfEmpties": {
    fields: [{ type: "fake_msgs/HasConstant", name: "arr", isArray: true }],
  },
  "fake_msgs/HasBigIntArrays": {
    fields: [{ type: "int64", name: "i_arr", isArray: true }, { type: "uint64", name: "u_arr", isArray: true }],
  },
  "fake_msgs/ContainsEverything": {
    fields: [
      { type: "std_msgs/Header", name: "first" },
      { type: "fake_msgs/HasComplexAndArray", name: "second" },
      { type: "fake_msgs/HasComplexArray", name: "third" },
      { type: "fake_msgs/HasConstant", name: "fourth" },
      { type: "fake_msgs/HasByteArray", name: "fifth" },
      { type: "fake_msgs/HasJson", name: "sixth" },
      { type: "fake_msgs/HasInt64s", name: "seventh" },
      { type: "fake_msgs/HasArrayOfEmpties", name: "eighth" },
      { type: "fake_msgs/HasBigIntArrays", name: "ninth" },
    ],
  },
};

export type HasComplexAndArray = $ReadOnly<{|
  header(): BinaryHeader,
  stringArray(): ArrayView<string>,
|}>;

export type HasComplexArray = $ReadOnly<{|
  complexArray(): ArrayView<HasComplexAndArray>,
|}>;

export type HasConstant = $ReadOnly<{|
  const(): number,
|}>;

export type HasByteArray = $ReadOnly<{|
  byte_array(): Uint8Array,
|}>;

export type HasJson = $ReadOnly<{|
  jsonField(): RosValue,
|}>;

export type HasInt64s = $ReadOnly<{|
  i64(bigInt: ?true): number,
  u64(bigInt: ?true): number,
|}>;

export type HasArrayOfEmpties = $ReadOnly<{|
  arr(): ArrayView<HasConstant>,
|}>;

export type HasBigIntArrays = $ReadOnly<{|
  i_arr(): ArrayView<number>,
  u_arr(): ArrayView<number>,
|}>;

export type ContainsEverything = $ReadOnly<{|
  first(): BinaryHeader,
  second(): HasComplexAndArray,
  third(): HasComplexArray,
  fourth(): HasConstant,
  fifth(): HasByteArray,
  sixth(): HasJson,
  seventh(): HasInt64s,
  eighth(): HasArrayOfEmpties,
  ninth(): HasBigIntArrays,
|}>;
