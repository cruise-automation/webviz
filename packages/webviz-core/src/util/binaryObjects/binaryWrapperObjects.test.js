// @flow
//
//  Copyright (c) 2020-present, Cruise LLC
//
//  This source code is licensed under the Apache License, Version 2.0,
//  found in the LICENSE file in the root directory of this source tree.
//  You may not use this file except in compliance with the License.

import { PointerExpression } from "./messageDefinitionUtils";
import { definitions } from "./testUtils";
import {
  printFieldDefinition,
  printGetClassForView,
  printSingularExpression,
} from "webviz-core/src/util/binaryObjects/binaryWrapperObjects";

describe("printSingularExpression", () => {
  it("handles strings", () => {
    const pointer = new PointerExpression("this.offset").add(1);
    expect(printSingularExpression({}, "string", pointer)).toBe(
      "$bigString.substr($view.getInt32((this.offset + 5), true), $view.getInt32((this.offset + 1), true))"
    );
  });

  it("handles JSON", () => {
    const pointer = new PointerExpression("this.offset").add(1);
    expect(printSingularExpression({}, "json", pointer)).toBe(
      "$context.parseJson($bigString.substr($view.getInt32((this.offset + 5), true), $view.getInt32((this.offset + 1), true)))"
    );
  });

  it("handles times", () => {
    const pointer = new PointerExpression("this.offset");
    expect(printSingularExpression(definitions, "time", pointer)).toBe("new time(this.offset)");
  });

  it("handles primitives", () => {
    const pointer = new PointerExpression("this.offset");
    expect(printSingularExpression(definitions, "int32", pointer)).toBe("$view.getInt32(this.offset, true)");
    expect(printSingularExpression(definitions, "uint64", pointer)).toBe("$int53.readUInt64LE($buffer, this.offset)");
    expect(printSingularExpression(definitions, "bool", pointer)).toBe("($view.getUint8(this.offset) !== 0)");
  });

  it("handles complex types", () => {
    const pointer = new PointerExpression("this.offset");
    expect(printSingularExpression(definitions, "std_msgs/Header", pointer)).toBe("new std_msgs_Header(this.offset)");
  });

  it("throws on unknown types", () => {
    const pointer = new PointerExpression("this.offset");
    expect(() => printSingularExpression(definitions, "not_a/Type", pointer)).toThrow("unknown type");
  });
});

describe("printFieldDefinition", () => {
  it("handles complex arrays", () => {
    const pointer = new PointerExpression("this.offset");
    const field = definitions["fake_msgs/HasComplexArray"].fields[0];
    expect(printFieldDefinition(definitions, field, pointer)).toBe(
      `
complexArray() {
  const from = $view.getInt32((this.offset + 4), true);
  const length = $view.getInt32(this.offset, true);
  return new fake_msgs_HasComplexAndArray$Array(from, length);
}`.trim()
    );
  });

  it("special-cases byte arrays", () => {
    const pointer = new PointerExpression("this.offset");
    const field = definitions["fake_msgs/HasByteArray"].fields[0];
    expect(printFieldDefinition(definitions, field, pointer)).toBe(
      `
byte_array() {
  const from = $view.getInt32((this.offset + 4), true);
  const length = $view.getInt32(this.offset, true);
  return new Uint8Array($arrayBuffer, from, length);
}`.trim()
    );
  });

  it("handles singular fields", () => {
    const pointer = new PointerExpression("this.offset");
    const field = definitions["std_msgs/Header"].fields[0];
    expect(printFieldDefinition(definitions, field, pointer)).toBe(
      `
seq() {
  return $view.getUint32(this.offset, true);
}`.trim()
    );
  });

  it("makes a method for constants", () => {
    const pointer = new PointerExpression("this.offset");
    const field = definitions["fake_msgs/HasConstant"].fields[0];
    expect(printFieldDefinition(definitions, field, pointer)).toBe(
      `
const() {
  return 1;
}`.trim()
    );
  });
});

describe("printGetClassForView", () => {
  it("returns the expected code", () => {
    expect(printGetClassForView(definitions, "fake_msgs/ContainsEverything", false)).toMatchSnapshot();
  });
});
