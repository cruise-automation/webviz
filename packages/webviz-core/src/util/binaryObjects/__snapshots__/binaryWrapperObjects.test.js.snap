// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`printGetClassForView returns the expected code 1`] = `
"const $offset = $context.offsetSymbol;
const $deepParse = $context.deepParse;
const $int53 = $context.int53;
const $arrayBuffer = $view.buffer;
const $buffer = $context.Buffer.from($arrayBuffer);
class fake_msgs_ContainsEverything {
  constructor(offset) {
    this[$offset] = offset;
  }
  first() {
    return new std_msgs_Header(this[$offset]);
  }
  second() {
    return new fake_msgs_HasComplexAndArray((this[$offset] + 20));
  }
  third() {
    return new fake_msgs_HasComplexArray((this[$offset] + 48));
  }
  fourth() {
    return new fake_msgs_HasConstant((this[$offset] + 56));
  }
  fifth() {
    return new fake_msgs_HasByteArray((this[$offset] + 56));
  }
  sixth() {
    return new fake_msgs_HasJson((this[$offset] + 64));
  }
  seventh() {
    return new fake_msgs_HasInt64s((this[$offset] + 72));
  }
  eighth() {
    return new fake_msgs_HasArrayOfEmpties((this[$offset] + 88));
  }
  ninth() {
    return new fake_msgs_HasBigIntArrays((this[$offset] + 96));
  }
  [$deepParse]() {
    return new deepParse$fake_msgs_ContainsEverything(this[$offset]);
  }
}
function deepParse$fake_msgs_ContainsEverything($offset) {
  this.first = new deepParse$std_msgs_Header($offset);
  this.second = new deepParse$fake_msgs_HasComplexAndArray(($offset + 20));
  this.third = new deepParse$fake_msgs_HasComplexArray(($offset + 48));
  this.fourth = new deepParse$fake_msgs_HasConstant(($offset + 56));
  this.fifth = new deepParse$fake_msgs_HasByteArray(($offset + 56));
  this.sixth = new deepParse$fake_msgs_HasJson(($offset + 64));
  this.seventh = new deepParse$fake_msgs_HasInt64s(($offset + 72));
  this.eighth = new deepParse$fake_msgs_HasArrayOfEmpties(($offset + 88));
  this.ninth = new deepParse$fake_msgs_HasBigIntArrays(($offset + 96));
}
$context.associateSourceData(fake_msgs_ContainsEverything, { datatypes: $typesByName, datatype: \\"fake_msgs/ContainsEverything\\", buffer: $arrayBuffer, bigString: $bigString })
class std_msgs_Header {
  constructor(offset) {
    this[$offset] = offset;
  }
  seq() {
    return $view.getUint32(this[$offset], true);
  }
  stamp() {
    return new time((this[$offset] + 4));
  }
  frame_id() {
    return $bigString.substr($view.getInt32((this[$offset] + 16), true), $view.getInt32((this[$offset] + 12), true));
  }
  [$deepParse]() {
    return new deepParse$std_msgs_Header(this[$offset]);
  }
}
function deepParse$std_msgs_Header($offset) {
  this.seq = $view.getUint32($offset, true);
  this.stamp = new deepParse$time(($offset + 4));
  this.frame_id = $bigString.substr($view.getInt32(($offset + 16), true), $view.getInt32(($offset + 12), true));
}
$context.associateSourceData(std_msgs_Header, { datatypes: $typesByName, datatype: \\"std_msgs/Header\\", buffer: $arrayBuffer, bigString: $bigString })
class fake_msgs_HasComplexAndArray {
  constructor(offset) {
    this[$offset] = offset;
  }
  header() {
    return new std_msgs_Header(this[$offset]);
  }
  stringArray() {
    const from = $view.getInt32((this[$offset] + 24), true);
    const length = $view.getInt32((this[$offset] + 20), true);
    return new string$Array(from, length);
  }
  [$deepParse]() {
    return new deepParse$fake_msgs_HasComplexAndArray(this[$offset]);
  }
}
function deepParse$fake_msgs_HasComplexAndArray($offset) {
  this.header = new deepParse$std_msgs_Header($offset);
  const stringArray$length = $view.getInt32(($offset + 20), true);
  const stringArray$from = $view.getInt32(($offset + 24), true);
  const stringArray$arr = new Array(stringArray$length);
  for (let $ptr = stringArray$from, $i = 0; $i < stringArray$length; $ptr += 8) {
    stringArray$arr[$i++] = $bigString.substr($view.getInt32(($ptr + 4), true), $view.getInt32($ptr, true));
  }
  this.stringArray = stringArray$arr;
}
$context.associateSourceData(fake_msgs_HasComplexAndArray, { datatypes: $typesByName, datatype: \\"fake_msgs/HasComplexAndArray\\", buffer: $arrayBuffer, bigString: $bigString })
class fake_msgs_HasComplexArray {
  constructor(offset) {
    this[$offset] = offset;
  }
  complexArray() {
    const from = $view.getInt32((this[$offset] + 4), true);
    const length = $view.getInt32(this[$offset], true);
    return new fake_msgs_HasComplexAndArray$Array(from, length);
  }
  [$deepParse]() {
    return new deepParse$fake_msgs_HasComplexArray(this[$offset]);
  }
}
function deepParse$fake_msgs_HasComplexArray($offset) {
  const complexArray$length = $view.getInt32($offset, true);
  const complexArray$from = $view.getInt32(($offset + 4), true);
  const complexArray$arr = new Array(complexArray$length);
  for (let $ptr = complexArray$from, $i = 0; $i < complexArray$length; $ptr += 28) {
    complexArray$arr[$i++] = new deepParse$fake_msgs_HasComplexAndArray($ptr);
  }
  this.complexArray = complexArray$arr;
}
$context.associateSourceData(fake_msgs_HasComplexArray, { datatypes: $typesByName, datatype: \\"fake_msgs/HasComplexArray\\", buffer: $arrayBuffer, bigString: $bigString })
class fake_msgs_HasConstant {
  constructor(offset) {
    this[$offset] = offset;
  }
  const() {
    return 1;
  }
  [$deepParse]() {
    return new deepParse$fake_msgs_HasConstant(this[$offset]);
  }
}
function deepParse$fake_msgs_HasConstant($offset) {

}
$context.associateSourceData(fake_msgs_HasConstant, { datatypes: $typesByName, datatype: \\"fake_msgs/HasConstant\\", buffer: $arrayBuffer, bigString: $bigString })
class fake_msgs_HasByteArray {
  constructor(offset) {
    this[$offset] = offset;
  }
  byte_array() {
    const from = $view.getInt32((this[$offset] + 4), true);
    const length = $view.getInt32(this[$offset], true);
    return new Uint8Array($arrayBuffer, from, length);
  }
  [$deepParse]() {
    return new deepParse$fake_msgs_HasByteArray(this[$offset]);
  }
}
function deepParse$fake_msgs_HasByteArray($offset) {
  const byte_array$length = $view.getInt32($offset, true);
  const byte_array$from = $view.getInt32(($offset + 4), true);
  this.byte_array = new Uint8Array($arrayBuffer, byte_array$from, byte_array$length);
}
$context.associateSourceData(fake_msgs_HasByteArray, { datatypes: $typesByName, datatype: \\"fake_msgs/HasByteArray\\", buffer: $arrayBuffer, bigString: $bigString })
class fake_msgs_HasJson {
  constructor(offset) {
    this[$offset] = offset;
  }
  jsonField() {
    return $context.parseJson($bigString.substr($view.getInt32((this[$offset] + 4), true), $view.getInt32(this[$offset], true)));
  }
  [$deepParse]() {
    return new deepParse$fake_msgs_HasJson(this[$offset]);
  }
}
function deepParse$fake_msgs_HasJson($offset) {
  this.jsonField = $context.parseJson($bigString.substr($view.getInt32(($offset + 4), true), $view.getInt32($offset, true)));
}
$context.associateSourceData(fake_msgs_HasJson, { datatypes: $typesByName, datatype: \\"fake_msgs/HasJson\\", buffer: $arrayBuffer, bigString: $bigString })
class fake_msgs_HasInt64s {
  constructor(offset) {
    this[$offset] = offset;
  }
  i64(bigInt) {
    return (bigInt ? $view.getBigInt64(this[$offset], true): $int53.readInt64LE($buffer, this[$offset]));
  }
  u64(bigInt) {
    return (bigInt ? $view.getBigUint64((this[$offset] + 8), true): $int53.readUInt64LE($buffer, (this[$offset] + 8)));
  }
  [$deepParse]() {
    return new deepParse$fake_msgs_HasInt64s(this[$offset]);
  }
}
function deepParse$fake_msgs_HasInt64s($offset) {
  this.i64 = $int53.readInt64LE($buffer, $offset);
  this.u64 = $int53.readUInt64LE($buffer, ($offset + 8));
}
$context.associateSourceData(fake_msgs_HasInt64s, { datatypes: $typesByName, datatype: \\"fake_msgs/HasInt64s\\", buffer: $arrayBuffer, bigString: $bigString })
class fake_msgs_HasArrayOfEmpties {
  constructor(offset) {
    this[$offset] = offset;
  }
  arr() {
    const from = $view.getInt32((this[$offset] + 4), true);
    const length = $view.getInt32(this[$offset], true);
    return new fake_msgs_HasConstant$Array(from, length);
  }
  [$deepParse]() {
    return new deepParse$fake_msgs_HasArrayOfEmpties(this[$offset]);
  }
}
function deepParse$fake_msgs_HasArrayOfEmpties($offset) {
  const arr$length = $view.getInt32($offset, true);
  const arr$from = $view.getInt32(($offset + 4), true);
  const arr$arr = new Array(arr$length);
  for (let $ptr = arr$from, $i = 0; $i < arr$length; $ptr += 0) {
    arr$arr[$i++] = new deepParse$fake_msgs_HasConstant($ptr);
  }
  this.arr = arr$arr;
}
$context.associateSourceData(fake_msgs_HasArrayOfEmpties, { datatypes: $typesByName, datatype: \\"fake_msgs/HasArrayOfEmpties\\", buffer: $arrayBuffer, bigString: $bigString })
class fake_msgs_HasBigIntArrays {
  constructor(offset) {
    this[$offset] = offset;
  }
  i_arr(bigInt) {
    const from = $view.getInt32((this[$offset] + 4), true);
    const length = $view.getInt32(this[$offset], true);
    return new int64$Array(from, length);
  }
  u_arr(bigInt) {
    const from = $view.getInt32((this[$offset] + 12), true);
    const length = $view.getInt32((this[$offset] + 8), true);
    return new uint64$Array(from, length);
  }
  [$deepParse]() {
    return new deepParse$fake_msgs_HasBigIntArrays(this[$offset]);
  }
}
function deepParse$fake_msgs_HasBigIntArrays($offset) {
  const i_arr$length = $view.getInt32($offset, true);
  const i_arr$from = $view.getInt32(($offset + 4), true);
  const i_arr$arr = new Array(i_arr$length);
  for (let $ptr = i_arr$from, $i = 0; $i < i_arr$length; $ptr += 8) {
    i_arr$arr[$i++] = $int53.readInt64LE($buffer, $ptr);
  }
  this.i_arr = i_arr$arr;
  const u_arr$length = $view.getInt32(($offset + 8), true);
  const u_arr$from = $view.getInt32(($offset + 12), true);
  const u_arr$arr = new Array(u_arr$length);
  for (let $ptr = u_arr$from, $i = 0; $i < u_arr$length; $ptr += 8) {
    u_arr$arr[$i++] = $int53.readUInt64LE($buffer, $ptr);
  }
  this.u_arr = u_arr$arr;
}
$context.associateSourceData(fake_msgs_HasBigIntArrays, { datatypes: $typesByName, datatype: \\"fake_msgs/HasBigIntArrays\\", buffer: $arrayBuffer, bigString: $bigString })
class time {
  constructor(offset) {
    this[$offset] = offset;
  }
  sec() {
    return $view.getInt32(this[$offset], true);
  }
  nsec() {
    return $view.getInt32((this[$offset] + 4), true);
  }
  [$deepParse]() {
    return new deepParse$time(this[$offset]);
  }
}
function deepParse$time($offset) {
  this.sec = $view.getInt32($offset, true);
  this.nsec = $view.getInt32(($offset + 4), true);
}
$context.associateSourceData(time, { datatypes: $typesByName, datatype: \\"time\\", buffer: $arrayBuffer, bigString: $bigString })
const string$Array = $context.getArrayView((offset) => $bigString.substr($view.getInt32((offset + 4), true), $view.getInt32(offset, true)), 8);
$context.associateSourceData(string$Array, { datatypes: $typesByName, datatype: \\"string\\", buffer: $arrayBuffer, bigString: $bigString, isArrayView: true });
const fake_msgs_HasComplexAndArray$Array = $context.getArrayView((offset) => new fake_msgs_HasComplexAndArray(offset), 28);
$context.associateSourceData(fake_msgs_HasComplexAndArray$Array, { datatypes: $typesByName, datatype: \\"fake_msgs/HasComplexAndArray\\", buffer: $arrayBuffer, bigString: $bigString, isArrayView: true });
const fake_msgs_HasConstant$Array = $context.getArrayView((offset) => new fake_msgs_HasConstant(offset), 0);
$context.associateSourceData(fake_msgs_HasConstant$Array, { datatypes: $typesByName, datatype: \\"fake_msgs/HasConstant\\", buffer: $arrayBuffer, bigString: $bigString, isArrayView: true });
const int64$Array = $context.getArrayView((offset) => $int53.readInt64LE($buffer, offset), 8, (offset) => $view.getBigInt64(offset, true));
$context.associateSourceData(int64$Array, { datatypes: $typesByName, datatype: \\"int64\\", buffer: $arrayBuffer, bigString: $bigString, isArrayView: true });
const uint64$Array = $context.getArrayView((offset) => $int53.readUInt64LE($buffer, offset), 8, (offset) => $view.getBigUint64(offset, true));
$context.associateSourceData(uint64$Array, { datatypes: $typesByName, datatype: \\"uint64\\", buffer: $arrayBuffer, bigString: $bigString, isArrayView: true });
return fake_msgs_ContainsEverything;"
`;
